cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowDefinitionValidator.java

```diff
@@ -0,0 +1,110 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.cps;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.codehaus.groovy.control.CompilationFailedException;
+import org.codehaus.groovy.control.MultipleCompilationErrorsException;
+import org.codehaus.groovy.control.messages.SyntaxErrorMessage;
+import org.codehaus.groovy.syntax.SyntaxException;
+
+import net.sf.json.JSONObject;
+
+public final class CpsFlowDefinitionValidator {
+
+    private static final String SUCCESS_KEY = "success";
+    private static final String FAIL_KEY = "fail";
+
+    private CpsFlowDefinitionValidator() {}
+
+    public static List<CheckStatus> toCheckStatus(CompilationFailedException x) {
+        List<CheckStatus> errors = new ArrayList<CheckStatus>();
+        if (x instanceof MultipleCompilationErrorsException) {
+            for (Object o : ((MultipleCompilationErrorsException) x).getErrorCollector().getErrors()) {
+                if (o instanceof SyntaxErrorMessage) {
+                    SyntaxException cause = ((SyntaxErrorMessage) o).getCause();
+                    CheckStatus st = new CheckStatus(cause.getOriginalMessage(), FAIL_KEY);
+                    st.setLine(cause.getLine());
+                    st.setColumn(cause.getStartColumn());
+                    errors.add(st);
+                }
+            }
+        }
+        if (errors.isEmpty()) {
+            // Fallback to simple message
+            CheckStatus st = new CheckStatus(x.getMessage(), FAIL_KEY);
+            st.setLine(1);
+            st.setColumn(0);
+            errors.add(st);
+        }
+        return errors;
+    }
+
+    public static final class CheckStatus {
+
+        public static final CheckStatus SUCCESS = new CheckStatus("", SUCCESS_KEY);
+
+        private String message;
+        private String status;
+        private int line;
+        private int column;
+
+        public CheckStatus(String message, String status) {
+            this.message = message;
+            this.status = status;
+        }
+        public String getMessage() {
+            return message;
+        }
+        public void setMessage(String message) {
+            this.message = message;
+        }
+        public String getStatus() {
+            return status;
+        }
+        public void setStatus(String status) {
+            this.status = status;
+        }
+        public int getLine() {
+            return line;
+        }
+        public void setLine(int line) {
+            this.line = line;
+        }
+        public int getColumn() {
+            return column;
+        }
+        public void setColumn(int column) {
```

Comment:
I followed the "only required fields in constructor" as in `@DataboundConstrucotr` (unrelated here, I know, but sensible). `column` is not required under some circumstances, so having a constructor with this field will lead in weird calls to `new CheckStatus("mesg", "success", 0, 0)` where `0, 0` does not really have any semantic. That's why I added this setters.